File: sprint3/wsdemoNoSTOMP/wsdemoNoSTOMP/src/main/java/it/unibo/wsdemoNoSTOMP/WebsocketClientEndpoint.java
package it.unibo.wsdemoNoSTOMP;

import javax.imageio.ImageIO;
import javax.websocket.*;
import java.awt.image.BufferedImage;
import java.io.ByteArrayInputStream;
import java.io.File;
import java.net.URI;
import java.nio.ByteBuffer;

@ClientEndpoint
public class WebsocketClientEndpoint {
    Session userSession = null; //initialized by the method onOpen
    private IMessageHandler messageHandler;

    public WebsocketClientEndpoint(URI endpointURI) {
        try {
            WebSocketContainer container=
                    ContainerProvider.getWebSocketContainer();
            container.connectToServer(this, endpointURI);
        } catch (Exception e) { throw new RuntimeException(e); }
    }

    //register message handler
    public void addMessageHandler(IMessageHandler msgHandler) {
        this.messageHandler = msgHandler;
    }

    //Send a message.
    public void sendMessage(String message) {
        this.userSession.getAsyncRemote().sendText(message);
    }

    @OnOpen
    public void onOpen(Session userSession) {
        this.userSession = userSession;
    }

    //Callback hook for Connection close events.
    @OnClose
    public void onClose(Session userSession, CloseReason reason) {
        this.userSession = null;
    }

    //Callback invoked when a client send a message.
    @OnMessage
    public void onMessage(String message) {
        if (this.messageHandler != null) {
            this.messageHandler.handleMessage(message);
        }
    }
    //Callback hook for images
    @OnMessage
    public void onMessage(ByteBuffer bytes) {
        try{
            ByteArrayInputStream bis = new ByteArrayInputStream(bytes.array());
            //Dai bytes alla immagine e salvataggio in un file
            BufferedImage bImage2    = ImageIO.read(bis);
            ImageIO.write(bImage2, "jpg", new File("outputimage.jpg") );
        }catch( Exception e){ throw new RuntimeException(e); }
    }

}

File: sprint3/WasteserviceStatusManager/src/main/kotlin/it/unibo/WasteserviceStatusManager/WebSocketClientEndpoint.kt
package it.unibo.WasteserviceStatusManager

import java.io.ByteArrayInputStream
import java.io.File
import java.net.URI
import java.nio.ByteBuffer
import javax.imageio.ImageIO
import javax.websocket.*

@ClientEndpoint
class WebSocketClientEndpoint(endpoint: URI) {
    private var userSession: Session? = null
    private lateinit var messageHandler: IMessageHandler

    init {
        try {
            val container = ContainerProvider.getWebSocketContainer()
            container.connectToServer(this, endpoint)
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }

    fun addMessageHandler(msgHandler: IMessageHandler) {
        this.messageHandler = msgHandler
    }

    fun sendMessage(message: String) {
        this.userSession?.asyncRemote?.sendText(message)
    }

    @OnOpen
    fun onOpen(userSession: Session) {
        this.userSession = userSession
    }

    @OnClose
    fun onClose(userSession: Session, reason: CloseReason) {
        this.userSession = null
    }

    @OnMessage
    fun onMessage(message: String) {
        this.messageHandler.handleMessage(message)
    }

    @OnMessage
    fun onMessage(bytes: ByteBuffer) {
        try {
            val bis = ByteArrayInputStream(bytes.array())
            val bImage2 = ImageIO.read(bis)
            ImageIO.write(bImage2, "jpg", File("outputimage.jpg"))
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }
}